design/sticky-session-design.md

```diff
@@ -0,0 +1,170 @@
+# Sticky session support
+
+**Status**: _draft_
+
+This document proposes adding support for sticky session in contour.
+
+## Goals
+
+- Enabling support for sticky session in Contour using
+    - Cookie
+    - Source IP
+    - Header
+- Adding support in IngressRoute CRD. 
+
+## Non-goals
+
+- Making options in IngressRoute very specific to Envoy is not a goal. Options in IngressRoute will be generic.
+
+## Background
+
+Sticky-sessions allow requests to be routed to the same host based on certain parameters of the request.
+The parameters can be a cookie, the source IP, a custom header, etc.
+
+Envoy supports sticky sessions by using a hash-based load balancing strategy (RingHash, Maglev). 
+Specifying a *hash policy* for such load-balancers will enable sticky sessions. However, we are not aiming at making
+this very specific to Envoy. IngressRoute options should be generic enough without exposing
+internals of Envoy itself.
+
+
+## High-Level Design
+
+Sticky sessions are enabled per route. The following change are required in IngressRoute:
+
+```yaml
+apiVersion: contour.heptio.com/v1beta1
+kind: IngressRoute
+metadata: 
+  name: basic
+  namespace: default
+spec: 
+  virtualhost:
+    fqdn: foo-basic.bar.com
+  routes: 
+    - match: /
+      sticky-session:
+        source: cookie/header/connection # There should be exactly one of these set.
+        name: name # This can be cookie name, header name, or connection property
+        ttl: time-to-live # This is only valid when source = cookie. 
+        path: cookie-path # This is only valid when source = cookie. 
+      services: 
+        - name: s1
+          port: 80  
+          strategy: RingHash # Optional. Default: RingHash
+```          
+
+For sticky sessions to be enabled on any route, the service associated to the route needs to have the 
+load balancing strategy set to a hash-based strategy (like RingHash or Maglev). The following cases need to be considered:
```

Comment:
If this is mandatory, then we shouldn't let users choose it. See my suggestion about `strategy: stick-session`.

Reply:
Seems right. I'm not sure if users have a strong preference for using one hash-based strategy against another. If that's not a concern, we could take care of that in the implementation and not have users to specify the hash type.

Secondly, in the context of [Envoy](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/load_balancing/load_balancers#ring-hash), the types RingHash and Maglev (which contour supports as a strategy) would behave as random without a hash policy. That's another thing to consider as well.